tests/language-feature/extensions/this-in-extension.slang

```diff
@@ -0,0 +1,38 @@
+// this-in-extension.slang
+
+// Test that an `This` type works correctly when there is an extension.
+
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECK): -shaderobj
+
+interface IFoo
+{
+    static const This identity;
+}
+
+__generic<T:IFoo>
+extension T
```

Comment:
This is little different from the repo example from the issue.
And even with this PR, we still cannot do the following, can we?
```
extension IFoo
```
It looks like `__generic<T:IFoo>` is a workaround for the limitation but I am not sure what the limitation exactly is.
In other words, why we cannot do `extension IFoo` here, instead?

Reply:
The whole purpose of the change is to get rid of `extension IFoo` from the language, because it does not make a lot of sense syntactically: you can't really add a requirement to an interface, because that will make all existing types conforming to the interface no longer conforming.

What `extension IFoo` really means is `extension<T:IFoo> T`, and this PR implements proper support for this, which will replace all existing uses of `extension IFoo`. The currently implementation of the compiler trying to support `extension IFoo` in a very hacky way that is not correct in many cases. It is much cleaner to just remove all these logic from the compiler and ban it forever.